### ISSUE TITLE
Unable to parallelize inner cores of GCE cluster

### ISSUE THREAD

**Issue by grantmcdermott** [cloudyr/googleComputeEngineR#129]
**Describe the bug**

This may well be an issue for `future.apply` rather than `googleComputeEngineR`, but let me try... 
I'm having trouble using all of the inner cores of a GCE cluster when using future.apply::future_apply(). The outer cores are parallelized (i.e. across the different VMs of the cluster), but each VM only runs in serial. I've read the [future topologies doc](https://cran.r-project.org/web/packages/future/vignettes/future-3-topologies.html), but I still can't seem to fix it. Since you provide an outline of several [related examples](https://cloudyr.github.io/googleComputeEngineR/articles/massive-parallel.html) online, I'm wondering if the fault is in how I create or call the cluster? (Alternatively, perhaps the examples in the docs suffer from this behaviour too?)

**To Reproduce**

Step 1. Set up a remote GCE cluster.
```
library(googleComputeEngineR)
## names for your cluster
vm_names <- c("testvm1","testvm2","testvm3")
## create the cluster using default template for r-base
## creates jobs that are creating VMs in background
jobs <- 
  lapply(
    vm_names, 
    function(x) {
      gce_vm(
        name = x,
        template = "r-base",
        predefined_type = "n1-highcpu-8",
        wait = FALSE
        )
      })
jobs
## wait for all the jobs to complete and VMs are ready
lapply(jobs, gce_wait)
# ## get the VM objects
vms <- lapply(vm_names, gce_vm)
## set up SSH for the VMs
vms <- lapply(vms, gce_ssh_setup)
vms
```
Step 2. Create a slow (bootstrapping) function that can be sped up in parallel
```
slow_func <- 
  function(i) {
    
    # Set sample size
    n <- 10000
    
    ## Generate a large data frame of fake data for a regression
    our_data <- data.frame(x = rnorm(n), e = rnorm(n))
    our_data$y = 3 + 2*our_data$x + our_data$e
    
    ## Sample the data
    sample_data <- our_data[sample.int(n = n, size = 1000, replace = T), ]
    ## Run the regression on our sampled data and extract the extract the x coefficient.
    x_coef <- lm(y ~ x, data = sample_data)$coef[2]
    ## Return value
    return(data.frame(x_coef = x_coef))
  }
```
Step 3. Run in parallel
```
library(future.apply)
## Set the plan
## See: https://cran.r-project.org/web/packages/future/vignettes/future-3-topologies.html
plan(list( 
  tweak(cluster, workers = as.cluster(vms)),  # Use a cluster of VMs
  tweak(multiprocess, workers = 8)  # Use all the CPUs on remote VMs 
  )) 
## Finally, run the function (takes about 13 seconds)
future_sim <- future_lapply(1:10000, slow_func) 
```
**Expected behavior**
I was expecting this function to run across all 3*8 = 24 cores on my cluster. However, SSH'ing in to the VMs and running `top` (and hitting "1") shows that only one core is being used per VM at a time. See screenshot below.

**Screenshots**
![top-gce](https://user-images.githubusercontent.com/10285037/53470648-1d323800-3a17-11e9-8b0c-43e438b47d99.png)

**Additional context**
I've left out the `sessionInfo()` since I doubt that will be informative (can always post if asked). Any ideas, though? Am I just expecting too much of `future_lapply` given the way my function is specified?

**Comment by MarkEdmondson1234** [comment]
Yes this is a limitation of this approach, so I only ever launch smaller VMs with 1-2 CPUs but more of them.  Perhaps there is some way to loop future within future calls, I'll see if @HenrikBengtsson can comment on an approach or raise an issue on the future repo

**Comment by grantmcdermott** [comment]
Thanks Mark.

@HenrikBengtsson you might be able to provide a quick solution, but I'd be happy to open this as a new issue for `future.apply` if you prefer.

**Comment by HenrikBengtsson** [comment]
In order to utilize the two-layered, 3-by-8 set of nested parallel workers, you need you use two layers of futures.  You could replace your one-layer futures:
```r
future_sim <- future_lapply(1:10000, slow_func)
```
with
```r
future_sim <- future_lapply(1:10000, FUN = function(x) {
  future_lapply(x, FUN = slow_func)
}, future.chunk.size = 10000/8)
```
And then do some post-shuffling to get the results into the proper format.

To parallelize on a single-layer of futures, you could launch 3-times-8 single-core machines, which has it's own pros and cons.

**Comment by grantmcdermott** [comment]
Thanks @HenrikBengtsson that's very helpful. 

I implemented your suggestion, first amending the future plan to install `future.apply` on each of the VMs:

```
plan(list( 
  # Use a cluster of computers locally 
  tweak(cluster, 
        workers = as.cluster(
          vms,
          ## Launch Rscript inside Docker container
          rscript = c(
            "docker", "run", "--net=host", "rocker/r-base",
            "Rscript"
          ),
          ## Install future.apply on each VM
          rscript_args = c("-e", shQuote("install.packages('future.apply')"))
          )
        ),  
  # Use all the CPUs on remote machines 
  tweak(multiprocess)  
  ))

## Now run the function
tic()
future_sim <- 
  future_lapply(1:10000, FUN = function(x) {
    future_lapply(x, FUN = slow_func)
  }, 
  future.chunk.size = 10000/8
  )
toc()
# 181.225 sec elapsed
```
The good news is that all cores on the VMs are now being used:

![top-recursive](https://user-images.githubusercontent.com/10285037/53510252-5ba91080-3a72-11e9-822e-eb6c289cf4d2.png)

The bad news is that overall run time is now much *slower* (around 3 minutes).  Am I doing something wrong, or is this just overhead outpacing the benefits of parallelization for this particular example? 

PS. One thing that makes me think it could be the former is that the CPU use across each VM is well below 100%.

**Comment by MarkEdmondson1234** [comment]
Very cool you can nest futures like that, I'd like to update the website with an example once you have a working one.  My suspicion is the overhead making things slower, is it possible to try see if a longer "big" function makes the figures look better?  It'll be good to have some figures to help work out the optimal structure.

**Comment by MarkEdmondson1234** [comment]
One modification I would mention would be to not have to install the future library on each VM.  You can use a custom Docker image that derives from https://github.com/rocker-org/rocker/issues/328

**Comment by grantmcdermott** [comment]
Sounds good, I was going to suggest adding an example like this to website too. 

FWIW, I'm still not convinced that it's an overhead problem. The fact that the individual cores are running so low makes me really suspicious that something's gone wrong. These always run at 100% when I parallelize this particular function on a single machine.

Still, I'm happy to close this issue now if you are.

PS - One more thing. I spotted a few more typos/errors in [the docs](https://cloudyr.github.io/googleComputeEngineR/articles/massive-parallel.html#create-the-cluster). E.g. The `lapply(vm_cluster, gce_get_zone_op)` function doesn't appear to exist it. I also tried `lapply(vm_cluster, gce_get_zone)` (which is documented) and a bunch of variants, but these all throw up errors. I can file a separate issue if you want, but may be an easy fix?

**Comment by MarkEdmondson1234** [comment]
Sure, if the underutilisation of CPU cores persists then that would be better raised with future issues.

It should be `gce_get_op` - will change it thanks!

**Comment by grantmcdermott** [comment]
Okay. Having played around with this some more tonight, I think I finally have an optimal solution or something close to it.

The trick is to manually split the outer and inner `future_lapply` calls. The outer call only loops over the number of VMs in the cluster (in this case: three). The inner call then loops over the desired number of iterations divided by the number of VMs in the cluster. A pretty simple cheat, but it seems to work. Here's the key code section. The `plan` stays the same as in my previous attempt https://github.com/cloudyr/googleComputeEngineR/issues/129#issuecomment-467956925.

```
## Desired number of iterations
iters <- 10000

## Now run the function
tic()
future_sim <- 
  future_lapply(1:length(vms), FUN = function(x) { ## Outer future_lapply() call loops over the no. of VMS
    future_lapply(1:(iters/length(vms)), FUN = slow_func) ## Inner future_lapply() call loops over desired no. of iterations / no. of VMs
  }
  )
toc()
## 3.738 sec elapsed (Much better!)
``` 

So, we're down to under four seconds, which is great. I'm not going to include the `top` screenshots this time, but these also confirm that all cores across the VM are running at maximum output.

A few things to note: 

- I couldn't get the nested iteration to optimize "automatically" using @HenrikBengtsson' suggestion of `future.chunk.size` (or, for that matter, `future.scheduling`). Maybe I'm doing something wrong. Interestingly, however, everything worked faster if I used  `future.chunk.size = 10000/3` instead of `future.chunk.size = 10000/8`... But still significantly slower than splitting the iteration inputs manually.
- I'm going to add the full working example to my course [lecture notes](https://github.com/uo-ec607/lectures#lecture-outline-and-quicklinks). @MarkEdmondson1234 let me know if you'd like me to add it to the official docs too, or if you'd prefer to handle that.
- Failing any objections, I think we can close this thread :)

**Comment by MarkEdmondson1234** [comment]
Thanks very much for taking the time to get to the bottom of it, would much welcome the docs updated with this example and will link to your tutorial as well to help people.  The website just needed to be built to update the docs above for the zone op, and I'd like to restructure the website a little to make it easier to find common workflows so that will be my next focus.
